### ISSUE TITLE
2.1 support

### ISSUE THREAD

**Issue by AlexFiliakov** [Glutanimate/cloze-overlapper#35]
Since Anki 2.0 support is ending in 2020.

**Comment by cverwey** [comment]
Any update on this? What are people using in the meanwhile as an alternative?

**Comment by jasonkena** [comment]
You can just build cloze-overlapper using anki-addon-builder, the instructions are on the `Readme.md` page

**Comment by woolfog** [comment]
When I try to build cloze-overlapper anki-addon-builder says "File not found - *.pyc".

**Comment by An-png** [comment]
Why is this tagged as announcement? Surely this should be high priority? Without 2.1 support, the majority of users cannot use this powerful addon

**Comment by An-png** [comment]
> When I try to build cloze-overlapper anki-addon-builder says "File not found - *.pyc".

I somehow successfully built it (on gnu/linux), but it doesn't work. An error comes up everytime you run anki.

The current version is not meant to work with anki 2.1, and it doesn't. The only solution is for it to be ported over.

**Comment by glutanimate** [comment]
This is a duplicate, please see #13 

As for building the add-on: It works fine for me with the latest version of `aab` from pypi. If you are experiencing issues building the add-on, I would appreciate it if you created a new issue on `aab`'s bug tracker: https://github.com/glutanimate/anki-addon-builder/issues. When doing so please make sure to include the full build log (terminal output) and the information on your system as requested in the bug report template. There's little I can do without that info.

@An-png Please create a new bug report with the error message you see when starting Anki, and please also include the information on your system, Anki version, etc. in the bug report template. Cloze Overlapper 0.4.0 works fine for me and many other users, so I need more info on your particular setup to know what's going on.

**Comment by An-png** [comment]
I got the build to work, it required **pyqt5-dev-tools**. This was not mentioned in the build instruction for this addon or the anki-addon-builder instructions. I realised this because the build needed "pyrcc5" and "pyuic5" . When typing "pyuic5" the terminal, i was told that "pygq5-dev-tools" needed to be installed so that pyuic5 can be used.

As for issue #13, it has been a year since you released that initial build. Why is it still limited to patreon? Even if you don't want to work on it, can't you release an alpha version on ankiweb addons like you did with image occulsion enhanced?

Sidenote: Why have you not released review heatmap & others on ankiweb even though they are freely available on github? There have been two copies of review heatmap on ankiweb already that didn't give credit to you. Instead of that, you could just release review heatmap, puppy reinforcement and other addons that are available on github onto ankiweb with (beta) listed in the title.

**Comment by glutanimate** [comment]
> I got the build to work, it required pyqt5-dev-tools. This was not mentioned in the build instruction for this addon or the anki-addon-builder instructions

Both tools are part of Anki's development environment, which `aab` presupposes: https://github.com/glutanimate/anki-addon-builder/#requirements

> As for issue #13, it has been a year since you released that initial build. Why is it still limited to patreon?

Because there are a number of core issues remaining to fix regarding persistent user data like the note type, card templates, and add-on configuration. Releasing at this point would mean more migration work later down the line. It makes more sense to get these core systems right before the add-on is circulating widely. Using Anki 2.0 for card creation and/or building CO from source remains a viable alternative for the time being.

> Why have you not released review heatmap & others on ankiweb even though they are freely available on github?

Review Heatmap, again, suffers from a number of core date/time related issues that are notoriously difficult to track down and fix because they are timezone-dependent and intertwined with issues that have only been recently (as in v2.1.20) fixed on Anki's side. It takes time to work on these.

AnkiWeb unfortunately does not offer a viable way to perform beta roll-outs (e.g. like the play store or app store do). If you upload a separate beta listing, there is no way to automatically direct users to the stable channel once beta testing is done. If the beta listing is gone, outside references linking to it will be broken. Maintaining an additional listing also takes more time and effort.

Worse still, people will abuse the review system for bug reports, and it's incredibly difficult to siphon out actionable debug info from users once they've posted a review. These reviews usually also persist regardless of whether the bug was addressed or not.

So while add-ons are in a usable, but still buggy state, I prefer for them to be downloadable through GitHub only. This encourages bug reports on the actual bug trackers, and usually provides me with much more workable feedback.

Ideally GitHub releases would lead by a few weeks ahead of AnkiWeb releases, and I think that period of time is a fairly reasonable balance between stability and update frequency. Unfortunately in the case of Review Heatmap and a few other add-ons, I had to deviate from that due to the sheer breadth of work needed to both port the add-ons to 2.1 and complete long-ongoing refactoring efforts. In retrospect it would have been better to simply push out a quick 2.1 port rather than attempting to do everything at once, but we're past that option.

Should completing the stable release of Review Heatmap take too much time still, then I will likely be going with a beta release on AnkiWeb. But I would really like to avoid that option as much as possible.

As for the other add-ons, I'm not sure which ones you're referring to, but Puppy Reinforcement for 2.1 has been available on AnkiWeb for a week now – after going through a thorough beta testing period that allowed me to fix quite a few remaining issues. Visual Feedback will be coming up next week, and more add-ons will follow soon.

**Comment by An-png** [comment]
Thanks to the clarification!
> Puppy Reinforcement for 2.1 has been available on AnkiWeb for a week now

Oh, I hadn't realised that. Last time I checked it wasn't on there. 

If you dislike adding alpha addons, why did you distrubute image occulsion enhanced as alpha? Is it because it was necessary in many people's anki workflow?

> Because there are a number of core issues remaining to fix regarding persistent user data like the note type, card templates, and add-on configuration. Releasing at this point would mean more migration work later down the line. It makes more sense to get these core systems right before the add-on is circulating widely. Using Anki 2.0 for card creation and/or building CO from source remains a viable alternative for the time being.

Considering Cloze Overlapper is one of your main addons ([even listed on your site](https://glutanimate.com/projects/)), why haven't you put any effort into it since November? Is it because it's very difficult to fix any problems with it?

**Comment by glutanimate** [comment]
> If you dislike adding alpha addons, why did you distrubute image occulsion enhanced as alpha? Is it because it was necessary in many people's anki workflow?

Yes, it's what I would consider an infrastructure add-on. It was also a mistake classifying it as an alpha. The only reason it has that designation is because a number of (fairly infrequently used) tools in the mask editor no longer work correctly on Anki's new web engine. To fix them I'll have to upgrade to a new version of the mask editor, which at the time seemed like something that would be simple to do (thus a quick step up to beta and then stable release on the horizon). Unfortunately it turned out to be a much tougher nut to crack than I anticipated.

So if I were to do things form the start, I would have just temporarily disabled all of these extra tools and shipped a stable build with a core feature-set, and then followed up on the missing features later on.

> Considering Cloze Overlapper is one of your main addons (even listed on your site), why haven't you put any effort into it since November? Is it because it's very difficult to fix any problems with it?

Just because the public commit log doesn't reflect any changes does not mean that the add-on has not seen any work. Some branches are not public, some things are stashed, and even more work is relegated to dependencies like [`libaddon`](https://github.com/glutanimate/anki-libaddon/), [`pytest-anki`](https://github.com/glutanimate/pytest-anki), or [Anki](https://github.com/ankitects/anki/commits?author=glutanimate) itself. Frankly, it's super myopic to make an assumption like that just by spending a few seconds looking at a commit log.

If you want to know what Anki add-on development truly feels like, [this is the best analogy I've found](https://www.youtube.com/watch?v=AbSehcT19u0). Not only are you constantly working on your add-ons, but you also have to work on building the ecosystem they run in. This is entirely different from more mature plugin ecosystems like the ones you find in web browsers or text editors.

When I started writing Anki add-ons there were no development tools available, the docs were sparse, and the existing add-on code I could study was ridden with hacks that were bound to break as soon as 2.0 moved out of the stasis it was in. All of that has improved tremendously since then, but we're still dealing with a young and ever-evolving platform. And to continue maintaining dozens of add-ons I had to do some groundwork that is not as fancy or publicly visible as a nice new feature or release.

For instance, Cloze Overlapper's development, more than that of any of my other add-ons, is invariably tied to `libaddon`'s progress, which in turn had to wait for `pytest-anki` to get to a workable state, which in turn was halted by Anki's rapidly evolving development pace over the last three months. 

Imagine writing a Chrome extension, but also having to help write the extension API, development tools, and Chrome itself at the same time. That's what add-on development is like for me at the moment, and that's why things take time.

If you only have to worry about one add-on, it's fairly easy to ship releases quickly. But without proper development tools and a more stable API, you simply cannot scale that to 70+ add-ons. I dedicated a lot of time these past couple of years towards achieving that scalability. That came at a cost in terms of release frequency and user-visible changes, but it was the only way for me to continue working on all of these add-ons going forward. Hopefully you will see all of that paying off soon.

**Comment by An-png** [comment]
> Just because the public commit log doesn't reflect any changes does not mean that the add-on has not seen any work. Some branches are not public, some things are stashed, and even more work is relegated to dependencies like libaddon, pytest-anki, or Anki itself. Frankly, it's super myopic to make an assumption like that just by spending a few seconds looking at a commit log.

Sorry! I assumed that was all the work you had put in to this addon because all the previous projects I had seen mainly did most of their work in the public commit log.

I think that this clarification was extremely useful, it might be useful to include this in a FAQ as to why some of your addons still haven't reached anki 2.1, [so that people don't start thinking this](https://github.com/glutanimate/cloze-overlapper/issues/13#issuecomment-555018633).

**Comment by glutanimate** [comment]
No worries, it's understandable. Communicating these types of things definitely is an area I need to improve on. It's just that it takes even more time away from coding :/

Still, an FAQ entry sounds like a good idea and I was meaning to update the docs soon anyway. So thank you for the suggestion!
